muxfs/mux.go

```diff
@@ -0,0 +1,167 @@
+// It's a way to use multiple filesystem implementations as a single FS
+type Mux struct {
+	handlers map[string]qfs.Filesystem
+}
+
+// compile-time assertion that MapStore satisfies the Filesystem interface
+var _ qfs.Filesystem = (*Mux)(nil)
+
+// SetHandler designates the resolver for a given path kind string
+func (m *Mux) SetHandler(pathKind string, resolver qfs.Filesystem) {
+	if m.handlers == nil {
+		m.handlers = map[string]qfs.Filesystem{}
+	}
+	m.handlers[pathKind] = resolver
+}
+
+// Option is a function that manipulates config details when fed to New(). Fields on
+// the o parameter may be null, functions cannot assume the Config is non-null.
+type Option func(o *[]MuxConfig) error
+
+// MuxConfig contains the information needed to create a new filesystem
+type MuxConfig struct {
+	Type   string                 `json:"type"`
+	Config map[string]interface{} `json:"options,omitempty"`
+	Source string                 `json:"source,omitempty"`
+}
+
+// constructors maps filesystem type strings to constructor functions
+var constructors = map[string]qfs.FSConstructor{
+	"ipfs":  qipfs.NewFilesystem,
+	"local": localfs.NewFilesystem,
+	"http":  httpfs.NewFilesystem,
+	"mem":   qfs.NewMemFilesystem,
+}
+
+// New creates a new Mux Filesystem, if no Option funcs are provided,
+// New uses a default set of Option funcs. Any Option functions passed to this
+// function must check whether their fields are nil or not.
+func New(ctx context.Context, cfgs []MuxConfig, opts ...Option) (*Mux, error) {
+	if cfgs == nil {
+		return nil, fmt.Errorf("config is required")
+	}
+
+	if len(opts) == 0 {
+		// default to a standard composition of Option funcs
+		opts = []Option{}
+	}
+
+	for _, opt := range opts {
+		if err := opt(&cfgs); err != nil {
+			return nil, err
+		}
+	}
+	mux := &Mux{
+		handlers: map[string]qfs.Filesystem{},
+	}
+	for _, cfg := range cfgs {
+		constructor, ok := constructors[cfg.Type]
+		if !ok {
+			return nil, fmt.Errorf("unrecognized filsystem type: %q", cfg.Type)
+		}
+		fs, err := constructor(cfg.Config)
+		if err != nil {
+			return nil, fmt.Errorf("constructing %q filesystem: %w", cfg.Type, err)
+		}
+		mux.handlers[cfg.Type] = fs
+	}
+	return mux, nil
+}
+
+func noMuxerError(kind, path string) error {
+	return fmt.Errorf("cannot resolve paths of kind '%s'. path: %s", kind, path)
+}
+
+// Get a path
+func (m Mux) Get(ctx context.Context, path string) (qfs.File, error) {
+	if path == "" {
+		return nil, qfs.ErrNotFound
+	}
+
+	kind := qfs.PathKind(path)
+	handler, ok := m.handlers[kind]
+	if !ok {
+		return nil, noMuxerError(kind, path)
+	}
+
+	return handler.Get(ctx, path)
+}
+
+// Put places a file or directory on the filesystem, returning the root path.
+// The returned path may or may not honor the path of the given file
+func (m Mux) Put(ctx context.Context, file qfs.File) (resPath string, err error) {
+	path := file.FullPath()
+	kind := qfs.PathKind(path)
```

Comment:
If I have an mux that contains ipfs and local (in order to load both types), and I want to put a file onto ipfs, does that mean I first need to set a path of "/ipfs/" in order to get the correct kind here? Or, alternatively, if I have a mux that *only* contains ipfs, do I still need to set this path of "/ipfs/" as to avoid the `noMuxerError`? Feels like there should be a optional fallback for `Put` that can be assigned and used if no path is set. That would be the common case of writing to ipfs a lot more convenience.

Reply:
Currently, yes, the `FullPath` would need to be `/ipfs/` if you are attempting to add it to ipfs.

However, using the `FullPath` as a metric for where the file should end up also feels a bit odd, as each implementation `Filestore` has its own implementations of `qfs.File`, and so has different expectations when you attempt to `Put` it into a filesystem. We probably should actually be switching on underlying `qfs.File` type and using that as a metric for where the file should end up. Or, having some ability to convert from on file type to another so that isn't as relevant, and we can rely on something else, whether it is the full path or a passed in param.

So yes, I agree that there needs to be more flexibility/thought surrounding `Put`. 

When working on `qfsutil` and attempting to copy one repo's contents into another this was one of the points that prevented me from getting very far.